src/libslic3r/GCode.cpp

```diff
@@ -3938,63 +3938,115 @@ std::string GCode::_before_extrude(const ExtrusionPath &path, const std::string
-            acceleration = std::min(max_acceleration, m_config.perimeter_acceleration.get_abs_value(acceleration));
-        } else if (m_config.bridge_acceleration.value > 0 && is_bridge(path.role())
-            && path.role() != erOverhangPerimeter) {
-            acceleration = std::min(max_acceleration, m_config.bridge_acceleration.get_abs_value(acceleration));
-        } else if (m_config.infill_acceleration.value > 0 && is_infill(path.role())) {
-            acceleration = std::min(max_acceleration, m_config.infill_acceleration.get_abs_value(acceleration));
+
+        if (m_config.travel_acceleration.value > 0)
+            travel_acceleration = m_config.travel_acceleration.get_abs_value(acceleration);
+
+        switch (path.role()){
+            case erPerimeter:
+                if (m_config.perimeter_acceleration.value >= 0)
+                    acceleration = m_config.perimeter_acceleration.get_abs_value(acceleration);
+                break;
+            case erExternalPerimeter:
+                if (m_config.external_perimeter_acceleration.value >= 0)
+                    acceleration = m_config.external_perimeter_acceleration.get_abs_value(acceleration);
+                break;
+            case erInternalInfill:
+                if (m_config.infill_acceleration.value >= 0)
+                    acceleration = m_config.infill_acceleration.get_abs_value(acceleration);
+                break;
+            case erSolidInfill:
+                if (m_config.solid_infill_acceleration.value >= 0)
+                    acceleration = m_config.solid_infill_acceleration.get_abs_value(acceleration);
+                break;
+            case erTopSolidInfill:
+            case erIroning:
+                if (m_config.top_solid_infill_acceleration.value >= 0)
+                    acceleration = m_config.top_solid_infill_acceleration.get_abs_value(acceleration);
+                break;
+            case erBridgeInfill:
+            case erInternalBridgeInfill:
+                if (m_config.bridge_acceleration.value >= 0)
+                    acceleration = m_config.bridge_acceleration.get_abs_value(acceleration);
+                break;
+            case erOverhangPerimeter:
+                if (m_config.overhangs_acceleration.value >= 0)
+                    acceleration = m_config.overhangs_acceleration.get_abs_value(acceleration);
+                break;
+            case erThinWall:
+            case erGapFill:
+            case erSkirt:
+            case erSupportMaterial:
+            case erSupportMaterialInterface:
+            case erWipeTower:
+            case erMilling:
+            case erCustom:
+            case erMixed:
+            case erCount:
+            case erNone:
+            default:
+                break;
         }
+
+        if (this->on_first_layer() && m_config.first_layer_acceleration.value > 0) {
+            acceleration = m_config.first_layer_acceleration.get_abs_value(acceleration);
```

Comment:
maybe it's better to use a min afterwards? If you have a 800 accel for the external perimeter, 2000 for perimeter and set 1200 for the first layer, you may don't want to increase your external perimeter accel. I don't think people want to increase it for the first layer?

Reply:
First layer always gets printed with the first layer acceleration, regardless of the extrusion role. If you set a percentage you can slow it down and the percentage will be relative to the respective extrusion role

Reply:
ok